libraries/AP_InertialSensor/AP_InertialSensor_RST.cpp

```diff
@@ -0,0 +1,401 @@
+#define GYRO_TEMP_OFFSET_CELSIUS        40
+
+
+// constructor
+AP_InertialSensor_RST::AP_InertialSensor_RST(AP_InertialSensor &imu,
+                              AP_HAL::OwnPtr<AP_HAL::SPIDevice> dev_gyro,
+                              AP_HAL::OwnPtr<AP_HAL::SPIDevice> dev_accel,
+                              enum Rotation rotation_g,
+                              enum Rotation rotation_a)
+    : AP_InertialSensor_Backend(imu)
+    , _dev_gyro(std::move(dev_gyro))
+    , _dev_accel(std::move(dev_accel))
+    , _rotation_g(rotation_g)
+    , _rotation_a(rotation_a)
+{
+}
+
+AP_InertialSensor_RST::~AP_InertialSensor_RST()
+{
+}
+
+/*
+  detect the sensor
+ */
+AP_InertialSensor_Backend *AP_InertialSensor_RST::probe(AP_InertialSensor &imu,
+                                              AP_HAL::OwnPtr<AP_HAL::SPIDevice> dev_gyro,
+                                              AP_HAL::OwnPtr<AP_HAL::SPIDevice> dev_accel,
+                                              enum Rotation rotation_g,
+                                              enum Rotation rotation_a)
+{
+    if (!dev_gyro && !dev_accel) {
+        return nullptr;
+    }
+    AP_InertialSensor_RST *sensor
+        = new AP_InertialSensor_RST(imu, std::move(dev_gyro), std::move(dev_accel), rotation_g, rotation_a);
+    if (!sensor || !sensor->_init_sensor()) {
+        delete sensor;
+        return nullptr;
+    }
+
+    return sensor;
+}
+
+/*
+ * init gyro
+ */
+bool AP_InertialSensor_RST::_init_gyro(void)
+{
+
+    _gyro_spi_sem = _dev_gyro->get_semaphore();
+
+    if (!_gyro_spi_sem->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
+        return false;
+    }
+
+    // set flag for reading registers
+    _dev_gyro->set_read_flag(0x80);
+
+    _dev_gyro->set_speed(AP_HAL::Device::SPEED_HIGH);
+
+    _dev_gyro->read_registers(GYRO_WHO_AM_I, &_gyro_whoami, sizeof(_gyro_whoami));
+    if (_gyro_whoami != GYRO_WHO_I_AM) {
+        hal.console->printf("RST: unexpected gyro WHOAMI 0x%x\n", (unsigned)_gyro_whoami);
+        printf("RST: unexpected gyro WHOAMI 0x%x\n", (unsigned)_gyro_whoami);
+        goto fail_whoami;
+    }
+
+    printf("detect i3g4250d\n");
+
+    //enter power-down mode first
+    _dev_gyro->write_register(GYRO_CTRL_REG1, 0);
+
+    hal.scheduler->delay(100);
+
+    _dev_gyro->write_register(GYRO_CTRL_REG1, 
+                              GYRO_REG1_POWER_NORMAL | GYRO_REG1_Z_ENABLE | GYRO_REG1_X_ENABLE | GYRO_REG1_Y_ENABLE | RATE_800HZ_LP_50HZ);
+
+    _dev_gyro->write_register(GYRO_CTRL_REG2, 0);        /* disable high-pass filters */
+    _dev_gyro->write_register(GYRO_CTRL_REG3, 0x0);        /* DRDY disable */
+    _dev_gyro->write_register(GYRO_CTRL_REG4, RANGE_2000DPS);
+    _dev_gyro->write_register(GYRO_CTRL_REG5, GYRO_REG5_FIFO_ENABLE);        /* disable wake-on-interrupt */
+
+    /* disable FIFO. This makes things simpler and ensures we
+     * aren't getting stale data. It means we must run the hrt
+     * callback fast enough to not miss data. */
```

Comment:
FIFO-enabled should be the ultimate goal as it allows to have higher frequency with lower CPU load.

Reply:
Thank you for your advice, I will test it in future, can we live with this now?

Reply:
yes, if it's working it's fine.